metric_learn/scml.py

```diff
@@ -0,0 +1,596 @@
+      wild be raised as this value will be needed.
+  gamma: float (default = 5e-3)
+      Learning rate
+  max_iter : int (default = 100000)
+      Number of iterations for the algorithm
+  output_iter : int (default = 5000)
+      Number of iterations to check current weights performance and output this
+      information in case verbose is True.
+  verbose : bool, optional
+      if True, prints information while learning
+  preprocessor : array-like, shape=(n_samples, n_features) or callable
+      The preprocessor to call to get triplets from indices. If array-like,
+      triplets will be formed like this: X[indices].
+  random_state : int or numpy.RandomState or None, optional (default=None)
+      A pseudo random number generator object or a seed for it if int.
+  Attributes
+  ----------
+  components_ : `numpy.ndarray`, shape=(n_features, n_features)
+      The linear transformation ``L`` deduced from the learned Mahalanobis
+      metric (See function `components_from_metric`.)
+  Examples
+  --------
+  >>> from metric_learn import SCLM_global_Supervised
+  >>> from sklearn.datasets import load_iris
+  >>> iris_data = load_iris()
+  >>> X = iris_data['data']
+  >>> Y = iris_data['target']
+  >>> scml = SCML_global_Supervised(basis='LDA', n_basis=400)
+  >>> scml.fit(X, Y)
+  References
+  ----------
+  .. [1] Y. Shi, A. Bellet and F. Sha. `Sparse Compositional Metric Learning.
+         <http://researchers.lille.inria.fr/abellet/papers/aaai14.pdf>`_. \
+         (AAAI), 2014.
+  .. [2] Adapted from original \
+         `Matlab implementation.<https://github.com/bellet/SCML>`_.
+  See Also
+  --------
+  metric_learn.SCML_global : The original weakly-supervised algorithm.
+
+  :ref:`supervised_version` : The section of the project documentation
+    that describes the supervised version of weakly supervised estimators.
+  """
+
+  def fit(self, triplets):
+    """Learn the SCML model.
+
+    Parameters
+    ----------
+    triplets : array-like, shape=(n_constraints, 3, n_features) or \
+          (n_constraints, 3)
+      3D array-like of triplets of points or 2D array of triplets of
+      indicators. In order to supervise the algorithm in the right way, we
+      should have the three samples ordered in a way such that:
+      d(triplets[i, 0],triplets[i, 1]) < d(triplets[i, 1], triplets[i, 3])
+      for all 0 <= i < n_constraints.
+
+    Returns
+    -------
+    self : object
+      Returns the instance.
+    """
+
+    return self._fit(triplets)
+
+
+class SCML_global_Supervised(_BaseSCML_global, TransformerMixin):
+  """Supervised version of Sparse Compositional Metric Learning (SCML)
+
+  `SCML_global_Supervised` creates triplets by taking `k_genuine` neighbours
+  of the same class and `k_impostor` neighbours from diferent classes for each
+  point and then runs the SCML algorithm on these triplets.
+  Read more in the :ref:`User Guide <scml>`.
+  Parameters
```

Comment:
Please use same line breaks convention as in docstring for other algorithms/functions
And most comments above apply here too

Reply:
+indent